xapian-applications/omega/clickmodel/tests/sdbntest.cc

```diff
@@ -0,0 +1,130 @@
+/** @file sdbntest.cc
+ * @brief tests for the Simplified DBN clickmodel class.
+ */
+/* Copyright (C) 2017 Vivek Pal
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+
+#include <cmath>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include <clickmodel/simplifieddbn.h>
+
+using namespace std;
+
+#define QID 0
+#define DOCIDS 1
+#define CLICKS 2
+
+struct sessions_items {
+    const string qid;
+    const string docids;
+    const string clicks;
+};
+
+struct sessions_testcase {
+    const string logfile;
+    const sessions_items sessions;
+};
+
+string sample_log1 = "test1.log";
+string sample_log2 = "test2.log";
+string sample_log3 = "test3.log";
```

Comment:
These strings should really be full system path to these files so no point in hardcoding their values. So, currently `sdbntest` will fail due to that reason. I think we need something like `test_driver::get_srcdir()` from `xapian-core/testsuite.h` here too or is there any other way to achieve the same goal?

Reply:
@ojwb @jaylett Ping?

Reply:
What do you think? Have you tried copying over the code from the test driver? What other way could work, looking at the code?

Reply:
I haven't tried the test driver code yet but it looks like it might work. But I'm not sure if copying over all the code just for using a single function is a wise thing to do so just wanted to ask if there might be another way of getting the directory path.

Reply:
Just copy the single function? It's not ideal, but it's probably a reasonable trade-off.

Reply:
@jaylett I tried but it doesn't work. Is there anything else I could try? Also, build errors on clang and OSX have been fixed with the latest commit. So, the only thing remaining is resolving the standing issue with getting the correct path for test files so that `sdbntest` pass.

Reply:
@ivmarkp I can't help you if you just say "it doesn't work". Obviously you cannot copy just the single function on its own, which is currently entangled in the rest of the test driver. However, you should be able to copy and adapt the code, resolving its dependencies; or write a new function that achieves the same thing.

Looking at the function, it has few dependencies beyond the STL. It's using `startswith()` and `endswith()`, plus `file_exists()`, all of which are in our common utils library (which is available in the omega build system, but you may have to link to get it).

Reply:
Actually, I did use the function correctly after resolving all its dependencies that you have mentioned but the problem was that the `srcdir` environment variable returns just `.` in
```c++
char *p = getenv("srcdir");
```
which should really return the path `"/home/checksum/Documents/xapian"` as found to be set in `configure` there generated by `./bootstrap`.

Reply:
@ivmarkp have a look at how the `runtest` script in xapian-core does this. You may need a similar approach (it's generated from `runtest.in`) to pass through `srcdir` for the omega tests.

Reply:
> which should really return the path `"/home/checksum/Documents/xapian"`

There's no need for the path to `srcdir` to be absolute (and in some ways a relative path is better - for example, it's less likely to exceed path name limits if you are building deep in a directory tree; it also makes it easier to compare the output of two builds in different directories).

The reason the tests fail currently is simple and nothing to do with `srcdir`.  The problem is simply that `./sdbntest` is run from the `omega` directory in the build tree, but tries to open `test1.log` (etc) with no path specified.  However, `test1.log` lives in `clickmodel/testdata/`.

If you prepend `clickmodel/testdata/` to the log filenames, then the tests pass when building in the source directory.  To fix this fully such that this also works when `$srcdir != $builddir`, you ought to prepend the value of `$srcdir` followed by `"/clickmodel/testdata/"`.

Reply:
@ojwb Great, thanks! Tests pass now.